url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/Utils/ConsoleWriter.java

```diff
@@ -0,0 +1,226 @@
+		StringBuffer sb = new StringBuffer();
+		sb.append("[ConsoleWriter]");
+		sb.append(": ");
+		sb.append(msg);
+		sb.append("\n");
+		if (e != null) {
+			sb.append(StringUtil.stackTraceString(e));
+		}
+
+		System.out.print(sb.toString());
+	}
+
+	/**
+	 * Console writer scheduler thread, calls sendingMessage() whenever a state change has been signaled.
+	 */
+	private class ConsoleWriterThread extends Thread {
+		private volatile boolean goOn = false;
+
+		private ConsoleWriterThread(String name) {
+			super(name);
+		}
+
+		@Override
+		public void run() {
+			goOn = true;
+
+			while (goOn) {
+				try {
+					stateChange.acquire();
+					while (sendingMessage()) {
+						;
+					}
+				} catch (InterruptedException e) {
+					// no action - expected when stopping or when deadline
+					// changed
+				} catch (Exception e) {
+					print("Unexpected exception caught in ConsoleWriter thread: ", e);
+				}
+			}
+		}
+
+		private void stopConsoleWriter() {
+			goOn = false;
+			this.interrupt();
+		}
+	}
+
+	/**
+	 * Start consolewriter scheduler thread. Should be called once at init time.
+	 */
+	public synchronized void startThread() {
+		print("Thread started.");
+		if (consoleWriterThread == null) {
+			consoleWriterThread = new ConsoleWriterThread("SERVER");
+			consoleWriterThread.start();
+		} else {
+			consoleWriterThread.interrupt();
+		}
+	}
+
+	/** Stop consolewriter scheduler thread. */
+	public void stopThread() {
+		print("Thread stopped");
+		if (consoleWriterThread != null) {
+			consoleWriterThread.stopConsoleWriter();
+			consoleWriterThread = null;
+		}
+	}
+
+	/**
+	 * Asks the server to identify this console. Puts the console code (generated by server) to a static variable to be
+	 * used among all classes.
+	 * 
+	 * A Console only needs to be started **once** per console!
+	 */
+	public void startConsole() {
+		if (consoleID.equals("")) {
+			consoleID = sendData("start=y");
+		}
+		System.out.println("===========================");
+		System.out.println("Console started: " + consoleID);
+		System.out.println("===========================");
+	}
+
+	/*
+	 * Messages
+	 */
+
+	/**
+	 * Called by agents in their print methods. Wraps the message around with source, and send with sendData()
+	 * 
+	 * @param sourceName
+	 *            agent's name
+	 * @param message
+	 *            the message to be sent
+	 */
+	public void sendMessage(String sourceName, String message) {
+		message = "[" + sourceName + "] " + message;
+		if (consoleID != "") {
+			message += "&consoleID=" + consoleID;
+		}
+		// System.out.println("Sending " + message);
+		messageQueue.add("message=" + message);
+		stateChanged();
+	}
+
+	/**
+	 * Alternatively, you can specify the source name by specifying in constructor, and pass only the message as param
+	 * when sending message.
+	 */
+	public void sendMessage(String message) {
+		if (name != null) {
+			message = "[" + name + "] " + message;
+		}
+		if (consoleID != "") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.